include/llvm/IR/DomTreeUpdater.h

```diff
@@ -0,0 +1,181 @@
+  /// \returns false when the class does not hold a Post dominator
+  /// tree or the Post dominator tree is not up to date.
+  bool hasPendingPostDomTreeUpdates() const;
+  // Modify
+  /// Apply updates according to the UpdateStrategy.
+  ///
+  /// \param Updates The updates to perform on data structures held.
+  void applyUpdates(ArrayRef<DominatorTree::UpdateType> Updates);
+  /// Insert an Edge based on UpdateStrategy.
+  ///
+  /// \param From The BasicBlock edge starts.
+  /// \param To The BasicBlock edge ends.
+  void insertEdge(BasicBlock *From, BasicBlock *To);
+  /// Delete an Edge based on UpdateStrategy.
+  ///
+  /// \param From The BasicBlock edge starts.
+  /// \param To The BasicBlock edge ends.
+  void deleteEdge(BasicBlock *From, BasicBlock *To);
+  /// Delete a basic block based on UpdateStrategy.
+  ///
+  /// \param DelBB The basic block to delete.
+  void deleteBB(BasicBlock *DelBB);
+  /// Delete a basic block based on UpdateStrategy with callback.
+  /// The callback will be called after DelBB -> removeFromParent()
+  /// and before delete DelBB; It will deduplicate with other DelBBs.
+  ///
+  /// \param DelBB The basic block to delete.
+  /// \param Callback The callback function to be called before delete DelBB;
+  void callbackDeleteBB(BasicBlock *DelBB, function_ref<void()> Callback);
+  /// Recalculate the dominator tree.
+  /// In this scenario, it is supposed the post dominator
+  /// tree should not be recalculated.
+  /// This function should be called when the class holds a DomTree.
+  void recalculateDomTree(Function &F);
+  /// Recalculate the post dominator tree.
+  /// In this scenario, it is supposed the dominator
+  /// tree should not be recalculated.
+  /// This function should be called when the class holds a PostDomTree.
+  void recalculatePostDomTree(Function &F);
+  /// Recalculate Both of the trees.
+  /// This function should be called when the class holds both of the trees.
+  void recalculateAll(Function &F);
+  // flush
+  /// FLush DomTree updates to get DomTree up to date.
+  /// It will also clear out of date updates and unnecessary
+  /// stored BasicBlocks.
+  /// It should only be called when there is a DomTree.
+  DominatorTree &flushDomTree();
+  /// FLush PostDomTree updates to get PostDomTree up to date.
+  /// It will also clear out of date updates and unnecessary
+  /// stored BasicBlocks.
+  /// It should only be called when there is a PostDomTree.
+  PostDominatorTree &flushPostDomTree();
+  /// Flush all pending updates to get DomTree and PostDomTree up to date.
+  /// PendingUpdates and PendingDeletedBBs will be cleared.
+  /// This function can be called not knowing which tree is held.
+  void flushAll();
+  // Debug
+  /// Debug method to help view the state of the internal data structures.
+  LLVM_DUMP_METHOD void dump() const;
+
+private:
+  class CallBackOnDeletion final : public CallbackVH {
+  public:
+    CallBackOnDeletion(Value *V, function_ref<void()> Callback)
```

Comment:
(I don't know if we need it, but)
Shouldn't this pass Value * as a parameter to callback? How can a callback know which node gets deleted?

Reply:
Previously, I just thought the caller can use lambdas to capture that Value*, which ignores the case that the API user just call it using a function.
I'll look further into this to see whether an overload is appropriate or just add Value* as a parameter to the callback.
```
void F(BasicBlock* BB){
  // Some codes
}
void N(){
  // Some codes
}
BasicBlock* B;
```
+ Original
```
callbackDeleteBB(B, [&B](){
  F(B);
});
callbackDeleteBB(B. N);
```
+ Proposed
```
callbackDeleteBB(B, F);
callbackDeleteBB(B, [&](BasicBlock* BB){
  N();
});
```

Reply:
Now I think there should be a BasicBlock* in the argument list in case the original BasicBlock* being modified later in the code.
e.g.
```
BasicBlock* B=(some A);
callbackDeleteBB(B, [&](){
  // use B
});
B = (some C);
DTU.flushAll();
// here the BB is actually deleted.
// But the callback affects (some C) which is not intended.
```
Or they should
```
BasicBlock* B=(some A);
callbackDeleteBB(B, [=B,some other stuffs](){
  // use B
});
B = (some C);
```